<img style="float: left;" src="earth-lab-logo-rgb.png" width="150" height="150">

# Earth Data Science Corps Summer 2020

![Colored Bar](colored-bar.png)

## Introduction to Using Spatial Vector Data in Open Source Python

<div class='notice--success' markdown="1">

## <i class="fa fa-ship" aria-hidden="true"></i> Fundamentals of Vector Data in Python 

In this lesson, you will learn fundamental concepts related to working with vector data in **Python**, including understanding the spatial attributes of vector data, how to open vector data and access its metadata.


## <i class="fa fa-graduation-cap" aria-hidden="true"></i> Learning Objectives

After completing this lesson, you will be able to:

* Describe the characteristics of 3 key vector data structures: points, lines and polygons.
* Open a shapefile in **Python** using **geopandas** - `gpd.read_file()`.
* View the CRS and other spatial metadata of a vector spatial layer in **Python**
* Access and view the attributes of a vector spatial layer in **Python**.

</div>

Add this reading as a to do for all of the info below so we are not duplicating information 
Reading: https://www.earthdatascience.org/courses/use-data-open-source-python/intro-vector-data-python/spatial-data-vector-shapefiles/ 

## About Spatial Vector Data

Vector data are composed of discrete geometric locations (x, y values) known as **vertices** that define the "shape" of the spatial object. The organization of the vertices determines the type of vector that you are working 
with. There are three types of vector data: 

* **Points:** Each individual point is defined by a single x, y coordinate. Examples of point data include: sampling locations, the location of individual trees or the location of plots.

* **Lines:** Lines are composed of many (at least 2) vertices, or points, that are connected. For instance, a road or a stream may be represented by a line. This line is composed of a series of segments, each "bend" in the road or stream represents a vertex that has defined `x, y` location.

* **Polygons:** A polygon consists of 3 or more vertices that are connected and "closed". Thus the outlines of plot boundaries, lakes, oceans, and states or countries are often represented by polygons. 
<figure>
    <a href="https://www.earthdatascience.org/images/earth-analytics/spatial-data/points-lines-polygons-vector-data-types.png">
    <img src="https://www.earthdatascience.org/images/earth-analytics/spatial-data/points-lines-polygons-vector-data-types.png" alt="There are 3 types of vector objects: points, lines or polygons. Each object type has a different structure. Image Source: Colin Williams (NEON)."></a>
    <figcaption> There are 3 types of vector objects: points, lines or polygons. Each object type has a different structure. Image Source: Colin Williams (NEON)
    </figcaption>
</figure>


## Introduction to the Shapefile Data Format Which Stores Points, Lines, and Polygons

Geospatial data in vector format are often stored in a `shapefile` 
format. Because the structure of points, lines, and polygons are 
different, each individual shapefile can only contain one vector 
type (all points, all lines or all polygons). You will not find 
a mixture of point, line and polygon objects in a single shapefile.

Objects stored in a shapefile often have a set of associated 
`attributes` that describe the data. For example, a line 
shapefile that contains the locations of streams, might 
contain the associated stream name, stream "order" and other 
information about each stream line object.

* More about shapefiles can found on 
<a href="https://en.wikipedia.org/wiki/Shapefile" target="_blank">Wikipedia</a>.


## Import Shapefiles

You will use the **geopandas** library to work with vector data in 
**Python**. You will also use `matplotlib.pyplot` to plot your data. 

In [1]:
# Import packages
import os
import matplotlib.pyplot as plt
import geopandas as gpd
import earthpy as et

# Get data and set working directory
data = et.data.get_data('spatial-vector-lidar')
os.chdir(os.path.join(et.io.HOME, 'earth-analytics'))

# TODO: 

https://www.naturalearthdata.com/downloads/50m-cultural-vectors/ 

lets use natural earth data instead here  - download the files using earthpy 
however but you can define each link for the students as a variable string

contries: https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/50m/cultural/ne_50m_admin_0_countries.zip
boundary lines - https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/50m/cultural/ne_50m_admin_0_boundary_lines_land.zip
populated places - https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/50m/cultural/ne_50m_populated_places_simple.zip

The shapefiles that you will import are:

* A polygon shapefile representing our field site boundary,
* A line shapefile representing roads, and
* A point shapefile representing the location of field sites at the <a href="http://www.neonscience.org/science-design/field-sites/harvard-forest" target="_blank"> San Joachin field site</a>.

The first shapefile that you will open contains the point locations of plots where trees have been measured. To import shapefiles you use the `geopandas` function `read_file()`. Notice that you call the `read_file()` function using `gpd.read_file()` to tell python to look for the function within the `geopandas` library.


In [2]:
# Define path to file
plot_centroid_path = os.path.join("data", "spatial-vector-lidar",
                                  "california", "neon-sjer-site",
                                  "vector_data", "SJER_plot_centroids.shp")

# Import shapefile using geopandas
sjer_plot_locations = gpd.read_file(plot_centroid_path)


## Spatial Data Attributes

Each object in a shapefile has one or more attributes associated with it.
Shapefile attributes are similar to fields or columns in a spreadsheet. Each row
in the spreadsheet has a set of columns associated with it that describe the row
element. In the case of a shapefile, each row represents a spatial object - for
example, a road, represented as a line in a line shapefile, will have one "row"
of attributes associated with it.

<figure>
    <a href="https://www.earthdatascience.org/images/earth-analytics/spatial-data/spatial-attribute-tables.png">
    <img src="https://www.earthdatascience.org/images/earth-analytics/spatial-data/spatial-attribute-tables.png" alt="A shapefile has an associated attribute table. Each spatial feature in a spatial object has the same set of
    associated attributes that describe or characterize the feature.
    Attribute data are stored in a separate .dbf file. Attribute data can be
    compared to a spreadsheet. Each row in a spreadsheet represents one feature
    in the spatial object. Image Source: National Ecological Observatory Network (NEON)"></a>
    <figcaption>Each spatial feature in a spatial object has the same set of
    associated attributes that describe or characterize the feature.
    Attribute data are stored in a separate *.dbf file. Attribute data can be
    compared to a spreadsheet. Each row in a spreadsheet represents one feature
    in the spatial object.
    Image Source: National Ecological Observatory Network (NEON)
    </figcaption>
</figure>


You can view the attribute table associated with our geopandas `GeoDataFrame` by typing the object name into the console (e.g., `sjer_plot_locations`). 

Or you can use the `.head(3)` method to display the first 3 rows of the attribute table. Adding a number fo the head method like this: `.head(6)` will specify how many rows of data python displays. 


In [3]:
# View top 6 rows of attribute table
sjer_plot_locations.head(6)

Plot_ID   Point     northing     easting plot_type  \
0  SJER1068  center  4111567.818  255852.376     trees   
1   SJER112  center  4111298.971  257406.967     trees   
2   SJER116  center  4110819.876  256838.760     grass   
3   SJER117  center  4108752.026  256176.947     trees   
4   SJER120  center  4110476.079  255968.372     grass   
5   SJER128  center  4111388.570  257078.867     trees   

                         geometry  
0  POINT (255852.376 4111567.818)  
1  POINT (257406.967 4111298.971)  
2  POINT (256838.760 4110819.876)  
3  POINT (256176.947 4108752.026)  
4  POINT (255968.372 4110476.079)  
5  POINT (257078.867 4111388.570)

In [4]:
# View the geometry type of each row
sjer_plot_locations.geom_type

0     Point
1     Point
2     Point
3     Point
4     Point
5     Point
6     Point
7     Point
8     Point
9     Point
10    Point
11    Point
12    Point
13    Point
14    Point
15    Point
16    Point
17    Point
dtype: object

<div class="notice--warning" markdown="1">

## <i class="fa fa-pencil-square-o" aria-hidden="true"></i> Challenge : Open Spatial Data in Python

Much like was done above, you're going to open your own spatial data with `geopandas`! Below we've provided a path to a dataset that contains spatial data similar to what we opened above. Open the spatial data with `geopandas`. Name the dataset `soap_plot_locations`. Once the dataset is opened, check the `geom_type`, and view the first few rows of the dataset using `dataset.head()`.

</div>

In [5]:
student_plot_centroid_path = os.path.join("data", "spatial-vector-lidar",
                                          "california", "neon-soap-site",
                                          "vector_data", "SOAP_centroids.shp")

# Open your dataset below this line. Make sure to view the geom_type and the first few rows of the dataset

soap_plot_locations = gpd.read_file(student_plot_centroid_path)

The cell below includes a set of tests to see if you correctly completed the activity in the cell above. They will provide you with feedback that can help you complete the activity. 

Be sure to run the cell below to check your code (please do not modify the cell!).

In [6]:
# Run this cell to ensure soap_plot_locations was created correctly

import notebook_tests_data_types

try:
    print(notebook_tests_data_types.test_geopandas_dataframe_creation(soap_plot_locations))
except NameError:
        print("'soap_plot_locations' is not defined. Make sure you spelled the variable name correctly!")


Variable 'soap_plot_locations' is a GeoDataFrame, good job!

'soap_plot_locations' has the correct amount of data, good job!


# TODO: consolidate this text and again reference the readings in the advanced textbook. because we have a 2 day workshop on spatial data i think we can just show them the basics here and get them plotting some data quickly. We will get into metadata in the more advanced workshop and can use those lessons then. 


i think we want to remove most fo what is below and just create some cool maps. there are other natural earth datasets that are small and could be used. let's have them create some quick maps of countries using data all int eh same CRS.

We will introduce more advanced crs stuff later.

Maybe at the bottom as a bonus or in the final activity we can have them perform some operation like maybe crop??

In this case, you have several attributes associated with your points including:

* Plot_ID, Point, easting, geometry, northing, plot_type 

<i class="fa fa-star"></i> **Data Tip:** The acronym, OGR, refers to the OpenGIS Simple Features Reference Implementation. <a href="https://trac.osgeo.org/gdal/wiki/FAQGeneral" target="_blank"> Learn more about OGR.</a>
{: .notice--warning }


### The Geopandas Data Structure

Notice that the geopandas data structure is a `dataframe` that contains a `geometry` column where the x, y point location values are stored. All of the other shapefile feature attributes are contained in columns, similar to what you may be used to if you've used a GIS tool such as ArcGIS or QGIS.



## Shapefile Metadata & Attributes

When you import the `SJER_plot_centroids` shapefile layer into `Python` the `gpd.read_file()` function automatically stores information about the data as attributes. You are particularly interested in the geospatial **metadata**, describing the format, `CRS`, `extent`, and other components of the vector data, and the **attributes** which describe properties associated with each individual vector object.


## Spatial Metadata

Key metadata for all shapefiles include:

1. **Object Type:** the class of the imported object.
2. **Coordinate Reference System (CRS):** the projection of the data.
3. **Extent:** the spatial extent (geographic area that the shapefile covers) of the shapefile. Note that the spatial extent for a shapefile represents the extent for ALL spatial objects in the shapefile.

You can view these shapefile metadata using the `.crs` and `.total_bounds` attributes:

In [7]:
# View object type
type(sjer_plot_locations)

geopandas.geodataframe.GeoDataFrame

In [8]:
# View CRS of object
sjer_plot_locations.crs

<Projected CRS: EPSG:32611>
Name: WGS 84 / UTM zone 11N
Axis Info [cartesian]:
- E[east]: Easting (metre)
- N[north]: Northing (metre)
Area of Use:
- name: World - N hemisphere - 120°W to 114°W - by country
- bounds: (-120.0, 0.0, -114.0, 84.0)
Coordinate Operation:
- name: UTM zone 11N
- method: Transverse Mercator
Datum: World Geodetic System 1984
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

The CRS for the data is epsg code: `32611`. You will learn about CRS formats and structures in a later lesson but for now a quick google search reveals that this CRS is: <a href="http://spatialreference.org/ref/epsg/wgs-84-utm-zone-11n/" target="_blank">UTM zone 11 North - WGS84</a>.

In [9]:
# View the spatial extent
sjer_plot_locations.total_bounds

array([ 254738.618, 4107527.074,  258497.102, 4112167.778])

<figure>
    <a href="https://www.earthdatascience.org/images/earth-analytics/spatial-data/spatial-extent.png">
    <img src="https://www.earthdatascience.org/images/earth-analytics/spatial-data/spatial-extent.png" alt="The spatial extent of a shapefile or geopandas GeoDataFrame represents the geographic edge or location that is the furthest north, south east and west. Thus is represents the overall geographic coverage of the spatial object. Image Source: National Ecological Observatory Network (NEON)."></a>
    <figcaption>The spatial extent of a shapefile or geopandas GeoDataFrame represents
    the geographic "edge" or location that is the furthest north, south east and
    west. Thus is represents the overall geographic coverage of the spatial object.
    Image Source: National Ecological Observatory Network (NEON)
    </figcaption>
</figure>

## How Many Features Are In Your Shapefile? 

You can view the number of features (counted by the number of rows in the attribute table) and feature attributes (number of columns) in our data using the pandas `.shape` method. Note that the data are returned as a vector of two values:

(rows, columns) 

Also note that the number of columns includes a column where the geometry (the x, y coordinate locations) are stored. 

In [10]:
sjer_plot_locations.shape

(18, 6)

<div class="notice--warning" markdown="1">

## <i class="fa fa-pencil-square-o" aria-hidden="true"></i> Challenge : Get Metadata from GeoPandas

Now you can get the same metadata, but for your `soap_plot_locations`. Assign the `crs` of your data to `soap_crs`, assign the bounds of your data to `soap_bounds`, and assign the shape of your data to `soap_shape`. You can `print()` those variables out to see how they differ from the metadata of the other dataset, and how they are similar as well! 

</div>

In [11]:
# Open your dataset below this line. Make sure to view the geom_type and the first few rows of the dataset

soap_crs = soap_plot_locations.crs

soap_bounds = soap_plot_locations.total_bounds

soap_shape = soap_plot_locations.shape


The cell below includes a set of tests to see if you correctly completed the activity in the cell above. They will provide you with feedback that can help you complete the activity. 

Be sure to run the cell below to check your code (please do not modify the cell!).

In [12]:
# Run this cell to ensure soap_crs, soap_bounds, and soap_shape were created correctly

try:
    print(notebook_tests_data_types.test_geopandas_dataframe_crs(soap_crs))
except NameError as e:
        print(e, "'soap_crs' is not defined. Make sure you spelled the variable name correctly!")
        
try:
    print(notebook_tests_data_types.test_geopandas_dataframe_bounds(soap_bounds))
except NameError:
        print("'soap_bounds' is not defined. Make sure you spelled the variable name correctly!")
        
try:
    print(notebook_tests_data_types.test_geopandas_dataframe_shape(soap_shape))
except NameError:
        print("'soap_shape' is not defined. Make sure you spelled the variable name correctly!")

Variable 'soap_crs' is a CRS object.
'soap_crs' has the correct projection data!
Variable 'soap_bounds' is a numpy array.
'soap_bounds' has the correct data!
Variable 'soap_shape' is a tuple.
'soap_shape' has the correct data!
